### PR TITLE
Implement Email and Password Validation

### PR DESCRIPTION
**Pull request by meshnesh**
## What this PR does
* Implement the feedback gotten for challenge 2
* Implement Email verification in user registration
* Implement Email and Password validation when login
* Implement Email verification when user RSVP to an Event
* Implement Tests on Event RSVP
* Implement Tests on email validation and verification in User Registration and Login

### CODE REVIEW

**Comment by betsybeth** [comment]
Follow conventions of creating a good PR.
Good work.

File: app/views.py [hunk 1]
```diff
@@ -1,35 +1,50 @@
...
+    'title': fields.String,
+    'location': fields.String,
+    'time': fields.String,
+    'date': fields.String,
+    'description': fields.String,
+    'done': fields.Boolean,
+    # 'uri': fields.Url('event')
+}
 
-class Event(Resource):
```
Comment (ann-mukundi) [inline app/views.py:30]: Follow pep8 standards, expected 2 two blank lines before the class.

File: app/views.py [hunk 2]
```diff
@@ -113,107 +99,143 @@ def put(self, event_id):
...
         }
-        return EVENTS[event_id], 201
+        EVENTS.append(event)
+        return {'event': marshal(event, EVENT_FIELDS)}, 201
 
-# Eventlist
-# shows a list of all events, and lets you POST to add new tasks
-
-class EventList(Resource):
+class Event(Resource):
```
Comment (ann-mukundi) [inline app/views.py:118]: Follow PEP 8 standards

File: app/views.py [hunk 3]
```diff
@@ -289,20 +327,28 @@ def post(self):
...
         """
         args = self.reqparse.parse_args()
-        USERS = {
+        users = {
             'email': args['email'],
             'password': args['password']
         }
-        return USERS, 200
+        for user in USERS:
+            print(user)
```
Comment (ann-mukundi) [inline app/views.py:346]: Remove print statement

File: app/views.py [hunk 4]
```diff
@@ -289,20 +327,28 @@ def post(self):
...
+            if users['email'] == user['email']:
+                if user['password'] == users['password']:
+                    return {'users': marshal(users, USER_LOGIN_FIELDS)}, 200
+                return 'Wrong password', 401
+        return 'Wrong email or email doesn\'t exist', 404
 
 # User reset password
 class PasswordRest(Resource):
```
Comment (ann-mukundi) [inline app/views.py:354]: PEP 8. Typo in classname.

File: test/tests.py [hunk 5]
```diff
@@ -21,96 +22,124 @@ def setUp(self):
...
         del self.event
         del self.test_client
+        del self.eventcreate
 
     def test_get_events(self):
         """Test that gets all events"""
-        res = self.event.get()
-        self.assertEqual(res, EVENTS)
+        res = self.test_client().get('/api/events')
+        self.assertEqual(res.status_code, 200)
```
Comment (ann-mukundi) [inline test/tests.py:30]: Code 200 does not necessarily imply that all events have been returned. For instance if you had two events and only one was returned the code would still be the same.

File: test/tests.py [hunk 6]
```diff
@@ -21,96 +22,124 @@ def setUp(self):
...
+        res = self.test_client().post('/api/events/1/rsvp', data=self.testrsvp)
+        self.assertEqual(res.status_code, 201)
+
+    def test_email_exist(self):
+        """Test user event rsvp"""
+        res = self.test_client().post('/api/events/1/rsvp', data=self.testemailexist)
+        self.assertEqual(res.status_code, 403)
+
+
 class TestUsers(unittest.TestCase):
```
Comment (ann-mukundi) [inline test/tests.py:84]: By lumping all your tests together you run the risk of having a test file that is too long. It is better to have each module's tests in a separate file.

File: test/tests.py [hunk 7]
```diff
@@ -21,96 +22,124 @@ def setUp(self):
...
         }
 
     def tearDown(self):
         del self.user
         del self.test_client
+        del self. createuser
+        del self.checkuser
 
-    def test_user_registration(self):
+    def test_user_reg_email_exist(self):
```
Comment (ann-mukundi) [inline test/tests.py:107]: I am having a hard time understanding what this is supposed to be testing. Is the goal to check whether a user can register with an already registered email? If thats the case a status code 404 does not seem right as the email actually exists.

File: test/tests.py [hunk 8]
```diff
@@ -21,96 +22,124 @@ def setUp(self):
...
         """Test user Registration"""
-        resp = self.test_client().post('/api/auth/register', data=self.createUser)
+        resp = self.test_client().post('/api/auth/register', data=self.checkuser)
+        self.assertEqual(resp.status_code, 404)
+
+    def test_reg_email_not_exist(self):
```
Comment (ann-mukundi) [inline test/tests.py:112]: Again here, the docstrings don't help me understand what you are testing as they are the same as those in <b> test_user_reg_email_exist </b>

File: test/tests.py [hunk 9]
```diff
@@ -21,96 +22,124 @@ def setUp(self):
...
+        """Test user Registration"""
+        resp = self.test_client().post('/api/auth/register', data=self.createuser)
         self.assertEqual(resp.status_code, 201)
 
     def test_user_login(self):
```
Comment (ann-mukundi) [inline test/tests.py:117]: It's great to see you test happy scenarios, I'd love to see more sad scenarios tested such as:
<ul>
<li>User login fails if required values are missing</li>
<li>User login fails if a n unregistered email is used</li>
<li>User login fail if wrong password is provided</li>
</ul>

**Review by ann-mukundi — CHANGES_REQUESTED** [review]
Please squash your commits the next time you do a PR.

**Pull request merged**